django/queryset/README.md

```diff
@@ -0,0 +1,340 @@
+        | Q(last_name__icontains=query)
+    )
+```
+
+First, we try to filter people with name `John`...
+
+```python
+people = find_people_yes(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+
+people = find_people_no(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+```
+
+Everything seems fine, right? Both techniques return the same results. But what about filtering person's name by `John Doe`?
+
+```python
+people = find_people_yes(query='John Doe')
+# return: <QuerySet [<Person: John Doe>]>
+
+people = find_people_no(query='John Doe')
+# return: <QuerySet []>
+```
+
+Now you see the difference? It's because the second method (`find_people_no`) does not support filtering across two fields.
+
+So if you have a requirement like this, you should join two fields together before filter them.
+
+### Using `.count()` instead of `len()`
+
+```python
+# Yes
+Person.objects.count()     # SELECT COUNT(*) FROM person
+
+# No
+len(Person.objects.all())  # SELECT * FROM person
+```
+
+Selecting aggregate function from database always consume less time that getting every records.
+
+### Using `.exists()` to check if any data exists
+
+```python
+# Yes
+if Person.objects.filter(first_name='John').exists():
+    do_something()
+
+# No
+if Person.objects.filter(first_name='John'):
+    do_something()
+```
+
+Using `.exists()` consume less resources than not using it.
+
+### Using `.first()` instead of calling `.exists()`, followed by `.get()`
+
+```python
+# Yes
+matched_person = Person.objects.filter(employee_id='123').first()
+if matched_person is not None:
+    do_something(matched_person)
+
+# No
+if Person.objects.filter(employee_id='123').exists():
+    matched_person = Person.objects.get(employee_id='123')
+    do_something(matched_person)
+```
+
+The reason is calling a single `.first()` will hit database only one. But when we call `.exists()`, it hits. Calling `.get()` will hits database again. Makes the second method to hit database twice which is a bad idea.
+
+### Avoid filtering across many relationship
+
+Filtering across _one-to-many_ or _many-to-many_ relationship can cause an unexpected result such as duplicated records.
+
+Suppose we have two departments
+
+```python
+hr = Department.objects.create(name='HRD')
+accounting = Department.objects.create(name='HRM')
+```
+
+John Doe is in both departments
+
+```python
+john_doe = Person.objects.create(...)
+hr.add(john_doe)
+accounting.add(john_doe)
+```
+
+If we try to get person who is in department that it's name contains word `HR`
+
+```python
+Person.objects.filter(departments__name__icontains='HR')
+# return: <QuerySet [<Person: John Doe>, <Person: John Doe>]>
+```
+
+We will get duplicated records because John Doe is in both departments. The solution for this case is to use `.distinct()`
+
+```python
+Person.objects.filter(departments__name__icontains='HR').distinct()
+# return: <QuerySet [<Person: John Doe>]>
```

Comment:
using `.distinct()` may hit database in terms of performance because it will compare every fields.  I suggest distinct only id column

`Person.objects.filter(departments__id ='HR').distinct('id')`

Reply:
`distinct(*fields)` will only work on PostgreSQL. 
In this case if you are using distinct `id` you must provide order by clause.

`Person.objects.filter(departments__id ='HR').order_by('id').distinct('id')`

Reply:
Fixed krub. Thanks krub P' ๆ